docs/two-phase-commit-transactions.md

```diff
@@ -500,6 +432,287 @@ public class ServiceBImpl implements ServiceB {
+      throws TransactionException {
+    // You can execute `prepare()` in parallel
+    List<TransactionException> exceptions =
+        Stream.of(transactions)
+            .parallel()
+            .map(
+                t -> {
+                  try {
+                    t.prepare();
+                    return null;
+                  } catch (TransactionException e) {
+                    return e;
+                  }
+                })
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+
+    // If any of the transactions failed to prepare, throw the exception
+    if (!exceptions.isEmpty()) {
+      throw exceptions.get(0);
+    }
+  }
+
+  private static void validate(TwoPhaseCommitTransaction... transactions)
+      throws TransactionException {
+    // You can execute `validate()` in parallel
+    List<TransactionException> exceptions =
+        Stream.of(transactions)
+            .parallel()
+            .map(
+                t -> {
+                  try {
+                    t.validate();
+                    return null;
+                  } catch (TransactionException e) {
+                    return e;
+                  }
+                })
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+
+    // If any of the transactions failed to validate, throw the exception
+    if (!exceptions.isEmpty()) {
+      throw exceptions.get(0);
+    }
+  }
+
+  private static void commit(TwoPhaseCommitTransaction... transactions)
+      throws TransactionException {
+    // You can execute `commit()` in parallel
+    List<TransactionException> exceptions =
+        Stream.of(transactions)
+            .parallel()
+            .map(
+                t -> {
+                  try {
+                    t.commit();
+                    return null;
+                  } catch (TransactionException e) {
+                    return e;
+                  }
+                })
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
+
+    // If any of the transactions succeeded to commit, you can regard the transaction as committed
+    if (exceptions.size() < transactions.length) {
+      if (!exceptions.isEmpty()) {
+        // You can log the exceptions here if you want
+      }
+
+      return; // Succeeded to commit
+    }
+
+    //
+    // If all the transactions failed to commit:
+    //
+
+    // If any of the transactions failed to commit due to `UnknownTransactionStatusException`, throw
+    // it because you should not retry the transaction in such a case
+    Optional<TransactionException> unknownTransactionStatusException =
+        exceptions.stream().filter(e -> e instanceof UnknownTransactionStatusException).findFirst();
+    if (unknownTransactionStatusException.isPresent()) {
+      throw unknownTransactionStatusException.get();
+    }
+
+    // Otherwise, throw the first exception
+    throw exceptions.get(0);
```

Comment:
[minor] This is a very corner case, but this line would throw NPE if `transactions` is empty?

Reply:
Ah, that's a very unexpected usage. We could add validation or an assertion to this method, but I think it might be too verbose since it's a sample code. What do you think?

Reply:
Yeah, I think it's okay to go with current simple one 👍